retrieval/target_provider.go

```diff
@@ -67,46 +78,79 @@ type sdTargetProvider struct {
 	refreshInterval time.Duration
 }
 
-// NewSdTargetProvider constructs a new sdTargetProvider for a job.
-func NewSdTargetProvider(job config.JobConfig) *sdTargetProvider {
-	i, err := utility.StringToDuration(job.GetSdRefreshInterval())
-	if err != nil {
-		panic(fmt.Sprintf("illegal refresh duration string %s: %s", job.GetSdRefreshInterval(), err))
-	}
+// NewSDTargetProvider constructs a new sdTargetProvider for a job.
+func NewSDTargetProvider(job config.JobConfig) TargetProvider {
 	return &sdTargetProvider{
 		job:             job,
-		refreshInterval: i,
+		refreshInterval: job.ServiceDiscoveryConfig().RefreshInterval(),
 	}
 }
 
-func (p *sdTargetProvider) Targets() ([]Target, error) {
-	var err error
+// Targets implements TargetPrvovider.
+func (p *sdTargetProvider) Targets() (targets []Target, err error) {
+	sd := p.job.ServiceDiscoveryConfig()
+	discoveryType := sd.GetType()
+
 	defer func() {
-		dnsSDLookupsCount.Inc()
+		sdLookupsCount.WithLabelValues(discoveryType).Inc()
 		if err != nil {
-			dnsSDLookupFailuresCount.Inc()
+			sdLookupFailuresCount.WithLabelValues(discoveryType).Inc()
 		}
 	}()
 
 	if time.Since(p.lastRefresh) < p.refreshInterval {
 		return p.targets, nil
 	}
+	baseLabels := clientmodel.LabelSet{
+		clientmodel.JobLabel: clientmodel.LabelValue(p.job.GetName()),
+	}
+	endpoint := &url.URL{
+		Scheme: "http://",
```

Comment:
I think this is about the point where we need to rethink the contents of the instance labels. I think it should just be a display name (usually host:port, but other things may make sense depending on the service discovery) and what we actually scrape is separate.